core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+        if (_active && !active) {
+            movePendingToActive();
+        }
+    }
+    _active = active;
+}
+
+
+/// Move all pending Tasks to the active heap.
+void ChunkTasks::movePendingToActive() {
+    for (auto const& t:_pendingTasks) {
+        LOGS(_log, LOG_LVL_DEBUG, "ChunkTasks " << _chunkId << " pending->active " << t->getIdStr());
+        _activeTasks.push(t);
+    }
+    _pendingTasks.clear();
+}
+
+
+/// @return true if active AND pending are empty.
+bool ChunkTasks::empty() const {
+    return _activeTasks.empty() && _pendingTasks.empty();
+}
+
+
+/// This is ready to advance when _activeTasks is empty and no Tasks are in flight.
+bool ChunkTasks::readyToAdvance() {
+    return _activeTasks.empty() && _inFlightTasks.empty();
+}
+
+
+/// @Return true if a Task is ready to be run.
+// ChunkTasks does not have its own mutex and depends on its owner for thread safety.
+// If a Task is ready to be run, _readyTask will not be nullptr.
+ChunkTasks::ReadyState ChunkTasks::ready(bool useFlexibleLock) {
+    auto logMemManRes =
+        [this](bool starved, std::string const& msg, std::vector<memman::TableInfo> const& tblVect) {
+            setResourceStarved(starved);
+            if (starved) {
+                std::string str;
+                for (auto const& tblInfo:tblVect) {
+                    str += tblInfo.tableName + " ";
+                }
+                LOGS(_log, LOG_LVL_DEBUG, "ready memMan " << msg << " - " << str);
+            }
+        };
+    if (_readyTask != nullptr) {
+        return ChunkTasks::ReadyState::READY;
+    }
+    if (_activeTasks.empty()) {
+        return ChunkTasks::ReadyState::NOT_READY;
+    }
+    // Calling this function doesn't get expensive until it gets here. Luckily,
+    // after this point it will return READY or NO_RESOURCES, and ChunkTaskList::_ready
+    // will not examine any further chunks upon seeing those results.
+    auto task = _activeTasks.top();
+    if (!task->hasMemHandle()) {
+        memman::TableInfo::LockType lckOptTbl = memman::TableInfo::LockType::MUSTLOCK;
+        memman::TableInfo::LockType lckOptIdx = memman::TableInfo::LockType::NOLOCK;
+        if (useFlexibleLock) lckOptTbl = memman::TableInfo::LockType::FLEXIBLE;
+        auto scanInfo = task-> getScanInfo();
+        auto chunkId = task->getChunkId();
+        if (chunkId != _chunkId) {
+            // This would slow things down badly, but the system would survive.
+            LOGS(_log, LOG_LVL_ERROR, "ChunkTasks " << _chunkId << " got task for chunk " << chunkId
+                    << " " << task->getIdStr());
+        }
+        std::vector<memman::TableInfo> tblVect;
+        for (auto const& tbl : scanInfo.infoTables) {
+            memman::TableInfo ti(tbl.db + "/" + tbl.table, lckOptTbl, lckOptIdx);
+            tblVect.push_back(ti);
+        }
+        // If tblVect is empty, we should get the empty handle
+        memman::MemMan::Handle handle = _memMan->lock(tblVect, chunkId);
```

Comment:
A key part of ready() is that it says a Task is "ready" to run. A critical part of that decision is that memMan has memory for the shared scan tables available, or they are already locked. The critical thing is that mmap has reserved space for the file. If all Tasks assume they'll be able to get mmap space, there will be dozens of tasks running, failing to get space and being scrubbed and added back to the scheduler, while another scheduler missed it's shot as all threads were used. It would be very difficult to make any order out of that.
 
One thing I have noticed, is that if mmap works, mlock always works as long as things are configured properly mmap is a fast call. The mlock call and mmap call could be split into separate functions. ready() would call mmap and the running Task would call mlock. 

The key part of the configuration is that the amount of memory MemMan believes it can lock is less than or equal to what the system will allow it to lock.

Conversely, in the other ticket DM-6518, if mlock fails, it is easy to pass the failure to the running Task via CommandMlock. Either way, the running Task knows mlock didn't work and appropriate action can be taken,
whatever that may be.